text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,288 @@
+We should implement a new tokenizer which detects script and tokenize based on it.
+
+### Prior Art and R&D
+
+**tokenization:**
+- > **[unicode-segmentation](https://github.com/unicode-rs/unicode-segmentation):**
+  > tokenizer which follow the [Standard Annex #29: Unicode Text Segmentation](http://www.unicode.org/reports/tr29/),
+  > this tokenizer seems promising for Latin scripts.
+- > **[Jieba](https://github.com/messense/jieba-rs):**
+  > tokenizer specialized in Chinese languages
+- > **[Lindera](https://github.com/lindera-morphology/lindera):**
+  > Japanese and Korean
+
+**lang/script detection:**
+- > **[whatlang](https://github.com/greyblake/whatlang-rs):**
+  > whatlang is able to detect script or/and language in a text,
+  > language detection is low but script is acceptable.
+  > note: Sonic also uses whatlang to peform the tokenization, it could be interesting to checkout how they do it.
+- > **toku (@qdequele):**
+  > in a R&D project, @qdequele was able to detect language based on stop word distribution in a text.
+  > If, in a latin script, there is lot of French stop words then the text language is probably french.
+
+**other solution that advertise multilingual support:**
+- Sonic uses whatlang to peform the tokenization, it could be interesting to checkout how they do it: https://github.com/valeriansaliou/sonic/tree/master/src/lexer
+	- Sonic uses whatlang to detect the languages, but don't acutally seems to use it to segment the text. It simply uses unicode segmentation, I can't really explain what they actually do with the language information.
+- tantivy advertise good multilingual support: https://github.com/tantivy-search/tantivy/tree/main/src/tokenizer
+	- Tantivy is similar to elastic in the sense that you can setup a custom text analyzer. The difference is that it is only made of `tokenizer` -> `token_filter`. Tantivy also provides a colleciton of tokenizer to choose from. Token are rather simple and do not contain any metadata, except for it's position.
+- How elastic search handle it: https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-tokenizers.html:
+	- Algolia proposed to create custom text analyzer. A text analyzer is a pipelined text processor with the collowing components: `char_filter` -> `tokenizer` -> `token_filter`. There are multiple diffferent tokenizer that can be chosed, depending on the usecase. This is a bit complicated, but I also think that advanced users should be able to chose the tokenizer they want. (default behaviour is us guessing what's the best tokenizer to use.)
+- Algolia:
+	- Algolia seems to use default word segmentation use user configurable separators. The use can also set the `queryLanguage` or provide a list of `stopWords`, but it is unclear how this work under the hood for languages such as Mandarin. They also word decompounging, meaning that they can split words that are mode of multiple words (mostly germanic/nordic languages apparently).
```

Comment:
They explain everything on the [doc](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/).

```suggestion
    - Algolia use multiple techniques to handle tokenization. They start by [normalize](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/normalization/) the data (lowercase, unidecode, transform traditional Chinese to modern, etc). Then, they tokenize the normalized data. It seems that the tokenization is based on two techniques. The first one is by [defining separators](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/in-depth/tokenization/) (space, coma, carriage return, etc) and the other one is with [dictionaries](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#using-dictionaries). The language is not automaticaly detected, [it must be set by the user](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/#no-automatic-language-detection). They also have dictionaries for plurials. 
```